source/extensions/filters/http/json_to_metadata/filter.cc

```diff
@@ -0,0 +1,376 @@
+      const auto& metadata = entry_it->second;
+      if (metadata.fields().contains(key)) {
+        ENVOY_LOG(trace, "Found key {} in namespace {}. Preserve the existing metadata value.", key,
+                  meta_namespace);
+        return false;
+      }
+    }
+  }
+
+  ENVOY_LOG(trace, "add metadata ns:{} key:{}", meta_namespace, key);
+
+  // Have we seen this namespace before?
+  auto namespace_iter = struct_map.find(meta_namespace);
+  if (namespace_iter == struct_map.end()) {
+    struct_map[meta_namespace] = ProtobufWkt::Struct();
+    namespace_iter = struct_map.find(meta_namespace);
+  }
+
+  auto& keyval = namespace_iter->second;
+  (*keyval.mutable_fields())[key] = std::move(val);
+
+  return true;
+}
+
+void Filter::finalizeDynamicMetadata(Http::StreamFilterCallbacks& filter_callback,
+                                     const StructMap& struct_map, bool& reported_flag) {
+  ASSERT(!reported_flag);
+  reported_flag = true;
+  if (!struct_map.empty()) {
+    for (auto const& entry : struct_map) {
+      filter_callback.streamInfo().setDynamicMetadata(entry.first, entry.second);
+    }
+  }
+}
+
+void Filter::handleAllOnMissing(const Rules& rules, bool& reported_flag) {
+  StructMap struct_map;
+  for (const auto& rule : rules) {
+    if (rule.rule().has_on_missing()) {
+      applyKeyValue(rule.rule().on_missing().value(), rule.rule().on_missing(), struct_map);
+    }
+  }
+
+  finalizeDynamicMetadata(*decoder_callbacks_, struct_map, reported_flag);
+}
+
+void Filter::handleOnMissing(const Rule& rule, StructMap& struct_map) {
+  if (rule.rule().has_on_missing()) {
+    applyKeyValue(rule.rule().on_missing().value(), rule.rule().on_missing(), struct_map);
+  }
+}
+
+void Filter::handleAllOnError(const Rules& rules, bool& reported_flag) {
+  StructMap struct_map;
+  for (const auto& rule : rules) {
+    if (rule.rule().has_on_error()) {
+      applyKeyValue(rule.rule().on_error().value(), rule.rule().on_error(), struct_map);
+    }
+  }
+  finalizeDynamicMetadata(*decoder_callbacks_, struct_map, reported_flag);
+}
+
+struct JsonValueToStringConverter {
+  std::string operator()(bool&& val) { return std::to_string(val); }
+  std::string operator()(int64_t&& val) { return std::to_string(val); }
+  std::string operator()(double&& val) { return std::to_string(val); }
+  std::string operator()(std::string&& val) { return std::move(val); }
+};
+
+struct JsonValueToDoubleConverter {
+  absl::StatusOr<double> operator()(bool&& val) { return static_cast<double>(val); }
+  absl::StatusOr<double> operator()(int64_t&& val) { return static_cast<double>(val); }
+  absl::StatusOr<double> operator()(double&& val) { return val; }
+  absl::StatusOr<double> operator()(std::string&& val) {
+    double dval;
+    if (absl::SimpleAtod(StringUtil::trim(val), &dval)) {
+      return dval;
+    }
+    return absl::InternalError(fmt::format("value {} to number conversion failed", val));
+  }
+};
+
+struct JsonValueToProtobufValueConverter {
```

Comment:
consider putting the various helper function objects in an anonymous namespace at the start of the file